vpr/src/place/cut_spreader.cpp

```diff
@@ -0,0 +1,857 @@
+                        || (radius > ripup_radius || rand() % (20000) < 10)) {
+                        ClusterBlockId bound = place_ctx.grid_blocks[sub_t.x][sub_t.y].blocks[sub_t.sub_tile];
+                        if (bound != EMPTY_BLOCK_ID) {
+                            if (imacro(bound) != -1) continue;
+                            unbindTile(sub_t); // else unbind sub_tile from bound
+                        }
+                        bindTile(sub_t, blk);
+                        if (iter_at_radius < need_to_explore) {
+                            unbindTile(sub_t);
+                            if (bound != EMPTY_BLOCK_ID) bindTile(sub_t, bound);
+                            int input_len = 0;
+                            for (auto pin : clb_nlist.block_input_pins(blk)) {
+                                ClusterNetId net = clb_nlist.pin_net(pin);
+                                if (net == ClusterNetId::INVALID() || clb_nlist.net_is_ignored(net) || clb_nlist.net_driver(net) == ClusterPinId::INVALID()) continue;
+                                ClusterBlockId driver = clb_nlist.pin_block(clb_nlist.net_driver(net));
+                                auto driver_loc = ap->blk_locs[driver].loc;
+                                input_len += std::abs(driver_loc.x - nx) + std::abs(driver_loc.y - ny);
+                            }
+                            if (input_len < best_inp_len) {
+                                best_inp_len = input_len;
+                                best_subtile = sub_t;
+                            }
+                            break;
+                        } else {
+                            if (bound != EMPTY_BLOCK_ID)
+                                remaining.emplace(imacro(bound) != -1 ? place_ctx.pl_macros[imacro(bound)].members.size() : 1, bound);
+                            ap->blk_locs[blk].loc = sub_t;
+                            placed = true;
+                            break;
+                        }
+                    }
+                }
+            } else {
+                for (auto sub_t : fast_t.at(nx).at(ny)) {
+                    std::vector<std::pair<ClusterBlockId, t_pl_loc>> targets;
+                    std::vector<std::pair<t_pl_loc, ClusterBlockId>> swaps_made;
+                    std::queue<std::pair<ClusterBlockId, t_pl_loc>> visit;
+                    visit.emplace(blk, sub_t);
+                    while (!visit.empty()) {
+                        ClusterBlockId vb = visit.front().first;
+                        VTR_ASSERT(!isPlaced(vb));
+                        t_pl_loc target = visit.front().second;
+                        visit.pop();
+                        auto log_type = clb_nlist.block_type(blk);
+                        auto result = std::find(log_type->equivalent_tiles.begin(), log_type->equivalent_tiles.end(), g_vpr_ctx.device().grid[target.x][target.y].type);
+                        if (result == log_type->equivalent_tiles.end()) goto fail;
```

Comment:
I don't like the goto here either.  Should make a cleanup routine and call it from two places and break if necessary.  Goto is almost never seen in C++ code nowadays.

I also prefer putting what an if controls in braces, or, if only one line you can do on the indented line under.  Putting on the same line is harder to read.  So:
if (result == log_type->equivalent_tiles.end()) {
    spreader_unlock ();   // Or some such
    continue;   // If you really need a continue
}



   
}